synapse/lib/grammar.py

```diff
@@ -0,0 +1,440 @@
+
+        assert kid.type == 'VARTOKN'
+        kids = self._convert_children(kids)
+        return s_ast.VarValue(kids)
+
+    def valulist(self, kids):
+        kids = self._convert_children(kids)
+        return s_ast.List(None, kids=kids)
+
+    def univpropvalu(self, kids):
+        kids = self._convert_children(kids)
+        return s_ast.UnivPropValue(kids=kids)
+
+    def switchcase(self, kids):
+        newkids = []
+
+        it = iter(kids)
+
+        varvalu = next(it)
+        newkids.append(varvalu)
+        assert isinstance(varvalu, s_ast.VarValue)
+
+        for casekid, sqkid in zip(it, it):
+            subquery = self._convert_child(sqkid)
+            if casekid.valu == '*':
+                caseentry = s_ast.CaseEntry(kids=[subquery])
+            else:
+                casekid = self._convert_child(casekid)
+                caseentry = s_ast.CaseEntry(kids=[casekid, subquery])
+
+            newkids.append(caseentry)
+
+        return s_ast.SwitchCase(newkids)
+
+    def casevalu(self, kids):
+        assert len(kids) == 1
+        kid = kids[0]
+
+        if kid.type == 'DOUBLEQUOTEDSTRING':
+            return self._convert_child(kid)
+
+        return s_ast.Const(kid.value[:-1])  # drop the trailing ':'
+
+
+with s_datfile.openDatFile('synapse.lib/storm.lark') as larkf:
+    _grammar = larkf.read().decode()
+
+QueryParser = lark.Lark(_grammar, start='query', propagate_positions=True)
+StormCmdParser = lark.Lark(_grammar, start='stormcmdargs', propagate_positions=True)
+
+class Parser:
+    '''
+    Storm query parser
+    '''
+    def __init__(self, text, offs=0):
+
+        self.offs = offs
+        assert text is not None
+        self.text = text.strip()
+        self.size = len(self.text)
+
+    def _larkToSynExc(self, e):
+        '''
+        Convert lark exception to synapse badGrammar exception
+        '''
+        mesg = regex.split('[\n!]', e.args[0])[0]
+        at = len(self.text)
+        if isinstance(e, lark.exceptions.UnexpectedCharacters):
+            mesg += f'.  Expecting one of: {", ".join(t for t in e.allowed)}'
+            at = e.pos_in_stream
+
+        return s_exc.BadSyntax(at=at, text=self.text, mesg=mesg)
+
+    def query(self):
+        '''
+        Parse the storm query
+
+        Returns (s_ast.Query):  instance of parsed query
+        '''
+        try:
+            tree = QueryParser.parse(self.text)
+        except lark.exceptions.LarkError as e:
+            raise self._larkToSynExc(e)
+        newtree = AstConverter(self.text).transform(tree)
+        newtree.text = self.text
+        return newtree
+
+    def stormcmdargs(self):
+        '''
+        Parse command args that might have storm queries as arguments
+        '''
+        try:
+            tree = StormCmdParser.parse(self.text)
+        except lark.exceptions.LarkError as e:
+            raise self._larkToSynExc(e)
+        newtree = AstConverter(self.text).transform(tree)
+        assert isinstance(newtree, s_ast.Const)
```

Comment:
if this is a hard requirement lets move it away from being a `assert` and make it an `if` so it can't be compiled away

Reply:
it is a logic error of catastrophic consequence if this fails